app/frontend/src/features/auth/jail/TOSSection.tsx

```diff
@@ -0,0 +1,34 @@
+import React, { useState } from "react";
+import Button from "../../../components/Button";
+import TOS from "../../../components/TOS";
+import { acceptTOS } from "../../../service/jail";
+
+export default function TOSSection({
+  updateJailed,
+}: {
+  updateJailed: () => void;
+}) {
+  const [completed, setCompleted] = useState(false);
+  const [loading, setLoading] = useState(false);
+
+  const accept = async () => {
+    setLoading(true);
+    const info = await acceptTOS();
+    if (!info.isJailed) {
+      updateJailed();
+    } else {
+      //if user is no longer jailed, this component will be unmounted anyway
+      setLoading(false);
+      setCompleted(true);
```

Comment:
I didn't get rid of the loading/completed state, as `acceptTOS` doesn't happen in the redux store, and it's possible when there is more than one Jail reason (like, your location is not set), that one section is completed but you aren't out of jail yet.

Reply:
With such complex async issues, I quickly move all domain logic and state to the model layer. The redux store in this case.